In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("car data.csv")

In [3]:
df.head()

Car_Name  Year  Selling_Price  Present_Price  Kms_Driven Fuel_Type   
0     ritz  2014           3.35           5.59       27000    Petrol  \
1      sx4  2013           4.75           9.54       43000    Diesel   
2     ciaz  2017           7.25           9.85        6900    Petrol   
3  wagon r  2011           2.85           4.15        5200    Petrol   
4    swift  2014           4.60           6.87       42450    Diesel   

  Seller_Type Transmission  Owner  
0      Dealer       Manual      0  
1      Dealer       Manual      0  
2      Dealer       Manual      0  
3      Dealer       Manual      0  
4      Dealer       Manual      0

In [4]:
#df.drop('Car_Name',axis=1,inplace=True)
df['Car_Name']=df['Car_Name'].str.split().str.slice(0,1).str.join(' ')
df.head()

Car_Name  Year  Selling_Price  Present_Price  Kms_Driven Fuel_Type   
0     ritz  2014           3.35           5.59       27000    Petrol  \
1      sx4  2013           4.75           9.54       43000    Diesel   
2     ciaz  2017           7.25           9.85        6900    Petrol   
3    wagon  2011           2.85           4.15        5200    Petrol   
4    swift  2014           4.60           6.87       42450    Diesel   

  Seller_Type Transmission  Owner  
0      Dealer       Manual      0  
1      Dealer       Manual      0  
2      Dealer       Manual      0  
3      Dealer       Manual      0  
4      Dealer       Manual      0

In [5]:
df.isnull().sum()

Car_Name         0
Year             0
Selling_Price    0
Present_Price    0
Kms_Driven       0
Fuel_Type        0
Seller_Type      0
Transmission     0
Owner            0
dtype: int64

In [6]:
df.duplicated().sum()

2

In [7]:
df.drop_duplicates(inplace=True,keep="first")

In [8]:
df=df.reset_index()
df=df.drop('index',axis=1)

In [9]:
df['age']  = 2023 - df['Year'] 
df.drop(['Year'], axis = 1, inplace= True)
df.reset_index()

index Car_Name  Selling_Price  Present_Price  Kms_Driven Fuel_Type   
0        0     ritz           3.35           5.59       27000    Petrol  \
1        1      sx4           4.75           9.54       43000    Diesel   
2        2     ciaz           7.25           9.85        6900    Petrol   
3        3    wagon           2.85           4.15        5200    Petrol   
4        4    swift           4.60           6.87       42450    Diesel   
..     ...      ...            ...            ...         ...       ...   
294    294     city           9.50          11.60       33988    Diesel   
295    295     brio           4.00           5.90       60000    Petrol   
296    296     city           3.35          11.00       87934    Petrol   
297    297     city          11.50          12.50        9000    Diesel   
298    298     brio           5.30           5.90        5464    Petrol   

    Seller_Type Transmission  Owner  age  
0        Dealer       Manual      0    9  
1        Dealer       Manual      0   10  
2        Dealer       Manual      0    6  
3        Dealer       Manual      0   12  
4        Dealer       Manual      0    9  
..          ...          ...    ...  ...  
294      Dealer       Manual      0    7  
295      Dealer       Manual      0    8  
296      Dealer       Manual      0   14  
297      Dealer       Manual      0    6  
298      Dealer       Manual      0    7  

[299 rows x 10 columns]

In [10]:
df.head()

Car_Name  Selling_Price  Present_Price  Kms_Driven Fuel_Type Seller_Type   
0     ritz           3.35           5.59       27000    Petrol      Dealer  \
1      sx4           4.75           9.54       43000    Diesel      Dealer   
2     ciaz           7.25           9.85        6900    Petrol      Dealer   
3    wagon           2.85           4.15        5200    Petrol      Dealer   
4    swift           4.60           6.87       42450    Diesel      Dealer   

  Transmission  Owner  age  
0       Manual      0    9  
1       Manual      0   10  
2       Manual      0    6  
3       Manual      0   12  
4       Manual      0    9

In [11]:
df['Car_Name'].unique()

array(['ritz', 'sx4', 'ciaz', 'wagon', 'swift', 'vitara', 's', 'alto',
       'ertiga', 'dzire', 'ignis', '800', 'baleno', 'omni', 'fortuner',
       'innova', 'corolla', 'etios', 'camry', 'land', 'Royal', 'UM',
       'KTM', 'Bajaj', 'Hyosung', 'Mahindra', 'Honda', 'Yamaha', 'TVS',
       'Hero', 'Activa', 'Suzuki', 'i20', 'grand', 'i10', 'eon', 'xcent',
       'elantra', 'creta', 'verna', 'city', 'brio', 'amaze', 'jazz'],
      dtype=object)

In [12]:
temp=['sx4',  'wagon','s','ignis', '800','corolla', 'camry', 'land','UM', 'grand', 'eon', 'xcent','elantra','city', 'brio', 'jazz']
temp2=[]
for i in range(0,len(df)):
    if df.iloc[i]['Car_Name'] in temp:
        temp2.append(i)
# len(temp2)
df=df.drop(temp2)
df.reset_index(inplace=True)
df.head()

index Car_Name  Selling_Price  Present_Price  Kms_Driven Fuel_Type   
0      0     ritz           3.35           5.59       27000    Petrol  \
1      2     ciaz           7.25           9.85        6900    Petrol   
2      4    swift           4.60           6.87       42450    Diesel   
3      5   vitara           9.25           9.83        2071    Diesel   
4      6     ciaz           6.75           8.12       18796    Petrol   

  Seller_Type Transmission  Owner  age  
0      Dealer       Manual      0    9  
1      Dealer       Manual      0    6  
2      Dealer       Manual      0    9  
3      Dealer       Manual      0    5  
4      Dealer       Manual      0    8

In [13]:
df.drop(columns='index',axis=1,inplace=True)

In [14]:
print(df["Fuel_Type"].unique())
print(df["Seller_Type"].unique())
print(df["Transmission"].unique())
print(df["Owner"].unique())

['Petrol' 'Diesel']
['Dealer' 'Individual']
['Manual' 'Automatic']
[0 1]


In [15]:
X = df.drop("Selling_Price",axis=1)
y = df["Selling_Price"]

In [16]:
X=pd.get_dummies(X,columns=["Car_Name","Fuel_Type","Seller_Type","Transmission","Owner"])

In [17]:
pd.set_option("display.max_columns",None)

X.head()

Present_Price  Kms_Driven  age  Car_Name_Activa  Car_Name_Bajaj   
0           5.59       27000    9            False           False  \
1           9.85        6900    6            False           False   
2           6.87       42450    9            False           False   
3           9.83        2071    5            False           False   
4           8.12       18796    8            False           False   

   Car_Name_Hero  Car_Name_Honda  Car_Name_Hyosung  Car_Name_KTM   
0          False           False             False         False  \
1          False           False             False         False   
2          False           False             False         False   
3          False           False             False         False   
4          False           False             False         False   

   Car_Name_Mahindra  Car_Name_Royal  Car_Name_Suzuki  Car_Name_TVS   
0              False           False            False         False  \
1              False           False            False         False   
2              False           False            False         False   
3              False           False            False         False   
4              False           False            False         False   

   Car_Name_Yamaha  Car_Name_alto  Car_Name_amaze  Car_Name_baleno   
0            False          False           False            False  \
1            False          False           False            False   
2            False          False           False            False   
3            False          False           False            False   
4            False          False           False            False   

   Car_Name_ciaz  Car_Name_creta  Car_Name_dzire  Car_Name_ertiga   
0          False           False           False            False  \
1           True           False           False            False   
2          False           False           False            False   
3          False           False           False            False   
4           True           False           False            False   

   Car_Name_etios  Car_Name_fortuner  Car_Name_i10  Car_Name_i20   
0           False              False         False         False  \
1           False              False         False         False   
2           False              False         False         False   
3           False              False         False         False   
4           False              False         False         False   

   Car_Name_innova  Car_Name_omni  Car_Name_ritz  Car_Name_swift   
0            False          False           True           False  \
1            False          False          False           False   
2            False          False          False            True   
3            False          False          False           False   
4            False          False          False           False   

   Car_Name_verna  Car_Name_vitara  Fuel_Type_Diesel  Fuel_Type_Petrol   
0           False            False             False              True  \
1           False            False             False              True   
2           False            False              True             False   
3           False             True              True             False   
4           False            False             False              True   

   Seller_Type_Dealer  Seller_Type_Individual  Transmission_Automatic   
0                True                   False                   False  \
1                True                   False                   False   
2                True                   False                   False   
3                True                   False                   False   
4                True                   False                   False   

   Transmission_Manual  Owner_0  Owner_1  
0                 True     True    False  
1                 True     True    False  
2                 True     True    False  
3                 True     True    F

In [18]:
X.columns

Index(['Present_Price', 'Kms_Driven', 'age', 'Car_Name_Activa',
       'Car_Name_Bajaj', 'Car_Name_Hero', 'Car_Name_Honda', 'Car_Name_Hyosung',
       'Car_Name_KTM', 'Car_Name_Mahindra', 'Car_Name_Royal',
       'Car_Name_Suzuki', 'Car_Name_TVS', 'Car_Name_Yamaha', 'Car_Name_alto',
       'Car_Name_amaze', 'Car_Name_baleno', 'Car_Name_ciaz', 'Car_Name_creta',
       'Car_Name_dzire', 'Car_Name_ertiga', 'Car_Name_etios',
       'Car_Name_fortuner', 'Car_Name_i10', 'Car_Name_i20', 'Car_Name_innova',
       'Car_Name_omni', 'Car_Name_ritz', 'Car_Name_swift', 'Car_Name_verna',
       'Car_Name_vitara', 'Fuel_Type_Diesel', 'Fuel_Type_Petrol',
       'Seller_Type_Dealer', 'Seller_Type_Individual',
       'Transmission_Automatic', 'Transmission_Manual', 'Owner_0', 'Owner_1'],
      dtype='object')

In [19]:
temp=['Car_Name_Activa',
       'Car_Name_Bajaj', 'Car_Name_Hero', 'Car_Name_Honda', 'Car_Name_Hyosung',
       'Car_Name_KTM', 'Car_Name_Mahindra', 'Car_Name_Royal',
       'Car_Name_Suzuki', 'Car_Name_TVS', 'Car_Name_Yamaha', 'Car_Name_alto',
       'Car_Name_amaze', 'Car_Name_baleno', 'Car_Name_ciaz', 'Car_Name_creta',
       'Car_Name_dzire', 'Car_Name_ertiga', 'Car_Name_etios',
       'Car_Name_fortuner', 'Car_Name_i10', 'Car_Name_i20', 'Car_Name_innova',
       'Car_Name_omni', 'Car_Name_ritz', 'Car_Name_swift', 'Car_Name_verna',
       'Car_Name_vitara', 'Fuel_Type_Diesel', 'Fuel_Type_Petrol',
       'Seller_Type_Dealer', 'Seller_Type_Individual',
       'Transmission_Automatic', 'Transmission_Manual', 'Owner_0', 'Owner_1']
for i in temp:
    X[i].replace({True:1,False:0},inplace=True)

In [20]:
X.head()

Present_Price  Kms_Driven  age  Car_Name_Activa  Car_Name_Bajaj   
0           5.59       27000    9                0               0  \
1           9.85        6900    6                0               0   
2           6.87       42450    9                0               0   
3           9.83        2071    5                0               0   
4           8.12       18796    8                0               0   

   Car_Name_Hero  Car_Name_Honda  Car_Name_Hyosung  Car_Name_KTM   
0              0               0                 0             0  \
1              0               0                 0             0   
2              0               0                 0             0   
3              0               0                 0             0   
4              0               0                 0             0   

   Car_Name_Mahindra  Car_Name_Royal  Car_Name_Suzuki  Car_Name_TVS   
0                  0               0                0             0  \
1                  0               0                0             0   
2                  0               0                0             0   
3                  0               0                0             0   
4                  0               0                0             0   

   Car_Name_Yamaha  Car_Name_alto  Car_Name_amaze  Car_Name_baleno   
0                0              0               0                0  \
1                0              0               0                0   
2                0              0               0                0   
3                0              0               0                0   
4                0              0               0                0   

   Car_Name_ciaz  Car_Name_creta  Car_Name_dzire  Car_Name_ertiga   
0              0               0               0                0  \
1              1               0               0                0   
2              0               0               0                0   
3              0               0               0                0   
4              1               0               0                0   

   Car_Name_etios  Car_Name_fortuner  Car_Name_i10  Car_Name_i20   
0               0                  0             0             0  \
1               0                  0             0             0   
2               0                  0             0             0   
3               0                  0             0             0   
4               0                  0             0             0   

   Car_Name_innova  Car_Name_omni  Car_Name_ritz  Car_Name_swift   
0                0              0              1               0  \
1                0              0              0               0   
2                0              0              0               1   
3                0              0              0               0   
4                0              0              0               0   

   Car_Name_verna  Car_Name_vitara  Fuel_Type_Diesel  Fuel_Type_Petrol   
0               0                0                 0                 1  \
1               0                0                 0                 1   
2               0                0                 1                 0   
3               0                1                 1                 0   
4               0                0                 0                 1   

   Seller_Type_Dealer  Seller_Type_Individual  Transmission_Automatic   
0                   1                       0                       0  \
1                   1                       0                       0   
2                   1                       0                       0   
3                   1                       0                       0   
4                   1                       0                       0   

   Transmission_Manual  Owner_0  Owner_1  
0                    1        1        0  
1                    1        1        0  
2                    1        1        0  
3                    1        1     

In [21]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 1)

In [22]:
from sklearn.preprocessing import StandardScaler,MinMaxScaler
temp=X_train.columns

sc=StandardScaler()
temp2=X_train
sc.fit(temp2)
X_train_sc=sc.transform(temp2)
X_test_sc=sc.transform(X_test)
X_train_sc=pd.DataFrame(X_train_sc,columns=temp)
X_test_sc=pd.DataFrame(X_test_sc,columns=temp)

mms=MinMaxScaler()
temp2=X_train
mms.fit(temp2)
X_train_mms=mms.transform(temp2)
X_test_mms=mms.transform(X_test)
X_train_mms=pd.DataFrame(X_train_mms,columns=temp)
X_test_mms=pd.DataFrame(X_test_mms,columns=temp)

multiple linear regression

In [23]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred = lr.predict(X_test)

lr_sc = LinearRegression()
lr_sc.fit(X_train_sc, y_train)
y_pred_sc = lr_sc.predict(X_test_sc)

lr_mms = LinearRegression()
lr_mms.fit(X_train_mms, y_train)
y_pred_mms = lr_mms.predict(X_test_mms)

#Accuracy
from sklearn.metrics import r2_score
print("Basic => ",r2_score(y_test, y_pred))
print("StandardScalar => ",r2_score(y_test, y_pred_sc))
print("MinMaxScalar => ",r2_score(y_test, y_pred_mms))

Basic =>  0.8694956876961344
StandardScalar =>  -7.196738454025327e+20
MinMaxScalar =>  0.8694956876961331


ridge

In [24]:
from sklearn.linear_model import Ridge
rd = Ridge()
rd.fit(X_train, y_train)
y_pred = rd.predict(X_test)

rd_sc = Ridge()
rd_sc.fit(X_train_sc, y_train)
y_pred_sc = rd_sc.predict(X_test_sc)

rd_mms = Ridge()
rd_mms.fit(X_train_mms, y_train)
y_pred_mms = rd_mms.predict(X_test_mms)

#Accuracy
from sklearn.metrics import r2_score
print("Basic => ",r2_score(y_test, y_pred))
print("StandardScalar => ",r2_score(y_test, y_pred_sc))
print("MinMaxScalar => ",r2_score(y_test, y_pred_mms))

Basic =>  0.934822561445174
StandardScalar =>  0.9223326611438286
MinMaxScalar =>  0.86401791701026


In [25]:
X_train.head(1)
columns=X_train.columns
np_arr=np.array([[0.57,18000,10,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,0,1,1,0]])
rd.predict(pd.DataFrame(np_arr,columns=columns))

array([0.32355717])

lasso

In [26]:
from sklearn.linear_model import Lasso
ls = Lasso()
ls.fit(X_train, y_train)
y_pred = ls.predict(X_test)

ls_sc = Lasso()
ls_sc.fit(X_train_sc, y_train)
y_pred_sc = ls_sc.predict(X_test_sc)

ls_mms = Lasso()
ls_mms.fit(X_train_mms, y_train)
y_pred_mms = ls_mms.predict(X_test_mms)

#Accuracy
from sklearn.metrics import r2_score
print("Basic => ",r2_score(y_test, y_pred))
print("StandardScalar => ",r2_score(y_test, y_pred_sc))
print("MinMaxScalar => ",r2_score(y_test, y_pred_mms))

Basic =>  0.8720787837033412
StandardScalar =>  0.7598445925573252
MinMaxScalar =>  0.3568582353253569


Decision Tree Regression

In [27]:
from sklearn.tree import DecisionTreeRegressor
temp=['absolute_error', 'friedman_mse', 'poisson', 'squared_error']
for i in temp:
    dtr = DecisionTreeRegressor(criterion=i,random_state = 5)
    dtr.fit(X_train, y_train)

    y_pred_dtr = dtr.predict(X_test)

    print(i," -> ",r2_score(y_test, y_pred_dtr))

# so poisson is best, r2 score=0.8375043587814766
dtr = DecisionTreeRegressor(criterion='poisson',random_state = 5)
dtr.fit(X_train, y_train)

y_pred_dtr = dtr.predict(X_test)

print(r2_score(y_test, y_pred_dtr))

absolute_error  ->  0.8443591659305809
friedman_mse  ->  0.7252047225554346
poisson  ->  0.8375043587814766
squared_error  ->  0.7245857621972163
0.8375043587814766


Random Forest Regression

In [28]:
from sklearn.ensemble import RandomForestRegressor
temp1=['absolute_error', 'friedman_mse', 'poisson', 'squared_error']
temp2=[100]
for i in range(200,601,200):
    temp2.append(i)

for i in temp1:
    for j in temp2:
        rf = RandomForestRegressor(n_estimators = j,criterion=i,random_state = 5)
        rf.fit(X_train, y_train)

        y_pred_rf = rf.predict(X_test)

        print(i," and ",j," -> ",r2_score(y_test, y_pred_rf))

# so we got great r2 score with friedman_mse  and  100  ->  0.8807247931506951
rf = RandomForestRegressor(n_estimators = j,criterion=i,random_state = 5)
rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)

r2_score(y_test, y_pred_rf)


absolute_error  and  100  ->  0.8848553411981193
absolute_error  and  200  ->  0.870766034946953
absolute_error  and  400  ->  0.8712484094824559
absolute_error  and  600  ->  0.8668444815273463
friedman_mse  and  100  ->  0.8807247931506951
friedman_mse  and  200  ->  0.8807437751603141
friedman_mse  and  400  ->  0.8822362458587113
friedman_mse  and  600  ->  0.8830190417077582
poisson  and  100  ->  0.8825952899725882
poisson  and  200  ->  0.8833009789156407
poisson  and  400  ->  0.879789278816347
poisson  and  600  ->  0.8771830013109642
squared_error  and  100  ->  0.8746167350090821
squared_error  and  200  ->  0.8786550067421245
squared_error  and  400  ->  0.882159216719326
squared_error  and  600  ->  0.8822835128506072


0.8822835128506072

K-Nearest Neighbor

In [29]:
from sklearn.neighbors import KNeighborsRegressor

for i in range(3,52,2):
    knn=KNeighborsRegressor(n_neighbors=i)
    knn.fit(X_train, y_train)

    y_pred_knn = knn.predict(X_test)

    print(i," -> ",r2_score(y_test, y_pred_knn))

# so knn works pretty bad here

3  ->  -0.12017765201673458
5  ->  -0.13044364139164988
7  ->  -0.19236993715063933
9  ->  -0.13680821762748496
11  ->  -0.1362293338749987
13  ->  -0.20116631060882972
15  ->  -0.19651719110014
17  ->  -0.18368002503778302
19  ->  -0.11100728764638035
21  ->  -0.08898607741189268
23  ->  -0.017099236038581545
25  ->  -0.039062076575021365
27  ->  -0.025323807400436493
29  ->  -0.03143985666183968
31  ->  -0.07192229885518042
33  ->  -0.08056366684564953
35  ->  -0.05083716016332507
37  ->  -0.04404307646658134
39  ->  -0.051761459816431765
41  ->  -0.062011549737877925
43  ->  -0.07455862667158386
45  ->  -0.024127535800984035
47  ->  -0.03509473713414235
49  ->  -0.03488856049585132
51  ->  -0.026681333744374536


So, the best we got is Ridge variant having r2 score=0.934822561445174

In [30]:
import joblib
joblib.dump(rd,'car_predict_save')
joblib.dump(df,"new_df_save")
joblib.dump(X_train,'train_df_save')

['train_df_save']

In [33]:
model=joblib.load('car_predict_save')
prediction=pd.DataFrame(np_arr,columns=columns)
model.predict(prediction)[0]

0.3235571709453744

In [32]:
df_new=joblib.load('new_df_save')
df_new.columns
df_new['Car_Name'].unique()
prediction['Car_Name_ritz']=1.0
prediction['Car_Name_ritz']

0    1.0
Name: Car_Name_ritz, dtype: float64